Rest API for car rental app

nico-bt opened the pull request:
Go to Readme: https://github.com/nico-bt/backend-car-rental-API/blob/development/README.md

### Live deploys to see UI and interaction:
- Frontend: https://frontend-car-rental-ivory.vercel.app/
- API: https://car-rental-api.adaptable.app/api/clients

(Obs: The free tier goes to sleep after a while, and the first load may take a few seconds.)

Pull Request Frontend:
https://github.com/nico-bt/frontend-car-rental/pull/1

nico-bt committed:
Create project with Nest cli. Basic structure

nico-bt committed:
Add .env to hide enviromental variables such as passwords and connections

nico-bt committed:
Add Prisma and DB settings

- Add Prisma as ORM
- Add car Model
- Add config and connect to a postgres DB cloud instance in Supabase with a connection string passed by .env

nico-bt committed:
Fix type error

nico-bt committed:
Add the car resource with CRUD operations, DTO, and validation

Created a new Nest.js module, controller, and service for the car resource.
Implemented CRUD (Create, Read, Update, Delete) operations for managing cars.
Added a Data Transfer Object (DTO) for input validation using class-validator.

nico-bt committed:
Fix error: check if element exists in db before updating

nico-bt committed:
Add Tests for car service

nico-bt committed:
Return the cars in descending order

Client wants to see in a table the list of cars in descending order of updated date.
Easy to retrieve them directly like that for render

nico-bt committed:
Add max limit to Km.

nico-bt committed:
Add CORS and PORT

- CORS for enabling access to browsers
- process.PORT variable for deployment

nico-bt committed:
Adjustments trying to deploy

For deploying on render.com and build on the platform,
@nest/cli is required to be a moved from devDependencies to dependencies

nico-bt committed:
Update year min/max

Client ask to restrain possible years to
2015 - 2023
to avoid mistakes entering data.

venutti commented on src/car/dto/update-car.dto.ts:
```diff
@@ -0,0 +1,51 @@
+  Min,
+} from 'class-validator';
+
+export class UpdateCarDto {
```
If the only difference between CreateCarDto and UpdateCarDto is the use of `@IsNotEmpty`/`@IsOptional`, i think that you can use a mapped-type instead https://docs.nestjs.com/openapi/mapped-types#partial
```suggestion
export class UpdateCarDto extends PartialType(CreateCarDto)  {}
```

nico-bt replied:
I didn't know about that feature, very practical.
I tested it locally and all the validations work as well.

venutti commented on src/car/dto/create-car.dto.ts:
```diff
@@ -0,0 +1,51 @@
+
+  @IsNotEmpty()
+  @IsInt()
+  @Min(2015)
+  @Max(2023)
```
You can store these magic numbers (and the following ones) in a `constants.ts` file

```ts
export const MIN_YEAR = 2015;
export const MAX_YEAR = 2023;
```

```suggestion
  @Min(MIN_YEAR)
  @Max(MAX_YEAR)
```

nico-bt replied:
You are right, that is more clear, also i think actual year should be dynamic
```
const MIN_YEAR=2015
const ACTUAL_YEAR= new Date().getFullYear()
```

venutti commented on src/car/car.service.ts:
```diff
@@ -0,0 +1,68 @@
+  constructor(private readonly prismaService: PrismaService) {}
+
+  async create(createCarDto: CreateCarDto): Promise<Car> {
+    const { marca, modelo, year, km, color, ac, pasajeros, cambios } =
+      createCarDto;
+    const newCar = await this.prismaService.car.create({
+      data: {
+        marca,
+        modelo,
+        year,
+        km,
+        color,
+        ac,
+        pasajeros,
+        cambios,
+      },
+    });
```
I think you could pass the createCarDto to the `data` property in the same way you do in the `update` method

```suggestion
    const newCar = await this.prismaService.car.create({
      data: createCarDto,
    });
```

nico-bt committed:
Add client model

nico-bt committed:
Add client resource

CRUD functionality for clients.

nico-bt committed:
Add soft delete

For regulatory reasons, items can't be deleted.
There has to be an historical list of all transactions.
The transaction item (to be created) will reference both car and client.

Obs: No hay en prisma un "paranoid" como muestra Fabricio con sequelize
en sqlite.
Por lo que agregué columna is_deleted a los modelos (default to false).
Y en la acción delete, en vez de borrar el recurso, cambio el valor de
is_deleted=true

nico-bt committed:
Add transaction model and props to car model

Add props to car model:
- price (forgot to add initially)
- is_rented to keep in sync with transactions to show in the ui
only available cars to is_rent

nico-bt committed:
Add transaction resource

Only adding transactions functionality for now

nico-bt committed:
Add is_renting to Client Model

Added property to get track if a client has a rented car.
Client can only has one car at time.
The app is not going to check calendar combinations and superpositions.

The transactions are recorded day to day, meaning that a car has
to return to place before it can be assign to a new client.

nico-bt committed:
Fix error in calculation of total price: missing parenthesis

nico-bt committed:
Return all transactions

nico-bt committed:
Include client and car nested in the transaction response, to show in report

nico-bt committed:
Check if client already has a rent in progress

nico-bt committed:
Add the funtionalities: Add, get and update to transaction resource

nico-bt committed:
Add constants

No more "magic numbers".
Added constants with info provided by the client.
Improve legibility

nico-bt committed:
Add order by rented

Client ask for default order, separating rented cars and updated date
of the car info

nico-bt committed:
Add default orderBy of the returned elements, asked by client to show

nico-bt committed:
Add finishTransaction and active query

Finish transaction to terminate rent contract when client returns the car.
It free the client and car to rent or be rented again in the system.

Added a query to retrieve all transactions to get the chance to
get only active transaction or all transactions (historial)

nico-bt committed:
Fix: add missing 'price' prop

nico-bt committed:
Add sort to separate by defect active and not active transactions

nico-bt committed:
Add car tests

nico-bt committed:
Add client tests

nico-bt committed:
Add transaction tests

nico-bt committed:
Add Swagger to make Docs

To make Swagger detecting the responses automatically I had to add
responses Dto's.
The types generated by Prisma did not work automatically with Swagger

nico-bt committed:
Fix not missing Prisma props

Swagger did not recognize enum types from Prisma schema automatically
@ApiProperty decorator should be added

FacundoCastillo12 commented:
Hi, I took a quick look at it and here are some recommendations. One would be not to use Spanish and English together in the code, what I would recommend is to decide for one. Personally, I would recommend using English for everything, as you have done in much of the code.

As for the tests I think they are fantastic and must be tremendous work. This more than a recommendation would be a suggestion for the future, if you want to save time you could try to create all the tests as end-to-end and test each endpoint from the controller with supertest as if you were actually using the application. I saw you did a little bit of that, but it could be extended to everything. You could do it using SQLite as a database that runs only in test. That way you wouldn't have to test the way you did. While it can be tricky to create that testing environment, once it is set up in your application it can help you reduce testing time in the long run. 

Another would be to give some order to the folders. I do not say to follow an architecture because it can be complicated, but to divide it in different folders to begin with can help to leave the code more organized. You could create a folder for the controller, one for the service inside each module and inside those folders you could create another test folder to put the separate tests.

nico-bt commented:
@FacundoCastillo12 gracias por el comment. 
Si, se me mezclaron los idiomas. Arranqué haciendo los modelos partiendo del .doc que hizo Fabricio con la consigna del ejercicio. Después seguí con todo en inglés y no cambié los modelos/forms...

I will look into your recommendation regarding end-to-end testing.
Thanks

nico-bt committed:
Add method to delete db for testing e2e database

nico-bt committed:
Delete console.log

nico-bt committed:
Add e2e testing

nico-bt committed:
Add local db instance with Docker

- Add a test db and a dev db with docker
- Add dotenv-cli to switch between .env.local and .env.test
- Add script in package.json to start local and test db
and selecting .env

nico-bt committed:
Fix script

nico-bt committed:
Update README.md

nico-bt committed:
Update README.md

typo

nico-bt committed:
Fix e2e testing script, add db port, add pre scripts

nico-bt committed:
Merge branch 'development' of https://github.com/nico-bt/backend-car-rental-API into development

nico-bt committed:
Update README.md

nico-bt committed:
Separate test-db and dev-db

Add scripts to separate the starting of the container.
So you can run test e2e without reseting the dev db
Before it was reinitializing everything with docker compose

nico-bt committed:
Merge branch 'development' of https://github.com/nico-bt/backend-car-rental-API into development

nico-bt committed:
Add image to readme

nico-bt committed:
Update README.md

nico-bt committed:
Update README.md

nico-bt committed:
update image

nico-bt committed:
Merge branch 'development' of https://github.com/nico-bt/backend-car-rental-API into development
